### PR TITLE
webserver: slow query logging with traceID

### PR DESCRIPTION
**Pull request by tsenart**
This changeset adds slow query logging with a traceID which should
help us jump from logs to traces when using Grafana Cloud, and find
trace ids of past requests that were slow.

### CODE REVIEW

File: web/trace.go [hunk 1]
```diff
@@ -45,8 +64,26 @@ func getTraceContext(ctx context.Context, opName string, traceEnabled bool, span
...
+
 func (s traceAwareSearcher) List(ctx context.Context, q query.Q, opts *zoekt.ListOptions) (*zoekt.RepoList, error) {
 	return s.Searcher.List(ctx, q, opts)
 }
 func (s traceAwareSearcher) Close()         { s.Searcher.Close() }
 func (s traceAwareSearcher) String() string { return s.Searcher.String() }
+
+func HTTPTraceMiddleware(next http.Handler) http.Handler {
+	tracer := opentracing.GlobalTracer()
+	return othttp.MiddlewareFunc(tracer, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (stefanhengl) [inline web/trace.go]:
```suggestion
	return othttp.MiddlewareFunc(tracer, func(w http.ResponseWriter, r *http.Request) {
```

Comment (stefanhengl) [reply web/trace.go]: nit: I think the conversion is not necesarry.

**Review by stefanhengl — APPROVED** [review]
LGTM

**Review by keegancsmith — COMMENTED** [review]
we already have request logging. Can't you add in the traceid to that rather than doubling the logs?

**Comment by tsenart** [comment]
> Can't you add in the traceid to that rather than doubling the logs?

Do you mean the tracedSearcher? Yes, we can add it there too actually. I'll make that change.

**Comment by tsenart** [comment]
No, actually we don't do any logging in `traceAwareSearcher`. Can you point out what you have in mind?

**Comment by keegancsmith** [comment]
> No, actually we don't do any logging in `traceAwareSearcher`.

traceAwareSearcher is something we added, I think we can do better / make whatever decision we want here that is better.

> Can you point out what you have in mind?

Look for `loggedSearcher`. It has access to the `ctx` so should be able to retrieve the traceid. Feel free to adjust this line to be json based/etc, whatever makes the most sense for loki.

File: cmd/zoekt-webserver/main.go [hunk 2]
```diff
@@ -424,11 +437,72 @@ func (s *loggedSearcher) StreamSearch(ctx context.Context, q query.Q, opts *zoek
...
+		return
 	}
-	log.Printf("DBUG: search q=%s Options{EstimateDocCount=%v Whole=%v ShardMaxMatchCount=%v TotalMaxMatchCount=%v ShardMaxImportantMatch=%v TotalMaxImportantMatch=%v MaxWallTime=%v MaxDocDisplayCount=%v} Stats{ContentBytesLoaded=%v IndexBytesLoaded=%v Crashes=%v Duration=%v FileCount=%v ShardFilesConsidered=%v FilesConsidered=%v FilesLoaded=%v FilesSkipped=%v ShardsSkipped=%v MatchCount=%v NgramMatches=%v Wait=%v}", q.String(), opts.EstimateDocCount, opts.Whole, opts.ShardMaxMatchCount, opts.TotalMaxMatchCount, opts.ShardMaxImportantMatch, opts.TotalMaxImportantMatch, opts.MaxWallTime, opts.MaxDocDisplayCount, stats.ContentBytesLoaded, stats.IndexBytesLoaded, stats.Crashes, stats.Duration, stats.FileCount, stats.ShardFilesConsidered, stats.FilesConsidered, stats.FilesLoaded, stats.FilesSkipped, stats.ShardsSkipped, stats.MatchCount, stats.NgramMatches, stats.Wait)
-	return err
+
+	if st == nil {
+		return
+	}
+
+	minDuration, _ := time.ParseDuration(os.Getenv("SRC_HTTP_LOG_MIN_DURATION"))
```
Comment (keegancsmith) [inline cmd/zoekt-webserver/main.go]: I find having every query logged quite useful for zoekt. It isn't as noisy as other logs. Can we either not include this or change the default to minDuration = 0

Also probably over thinking it, but would be great to just parse this once? It's essentially a global so might as well parse as a global? Or parse in the duration as a field to loggedSearcher. Or finally, just remove this feature :)

**Review by keegancsmith — APPROVED** [review]
shipit!

minor/can-ignore: A bunch of unrelated formatting changes to files not changed. I guess its from the earlier version fo this PR?

File: cmd/zoekt-webserver/main.go (hunk 2 above)
Comment (tsenart) [reply cmd/zoekt-webserver/main.go]: > Or finally, just remove this feature :)

Will go for this one if you don't think logging every query is a problem :)

**Pull request merged**